src/lib.rs

```diff
@@ -201,6 +201,10 @@ impl XTCTrajectory {
 impl Trajectory for XTCTrajectory {
     fn read(&mut self, frame: &mut Frame) -> Result<()> {
         let mut step: i32 = 0;
+        frame.resize(
```

Comment:
I don't think the xtc or trr specification allows individual frames to differ in the number of atoms.
Checking this for every read could be a big performance loss when iterating over trajectories. I think we dont need this and users should be aware that it's their responsibility to have a correctly sized Frame (and a trajectory following the specification).

Reply:
I think we need to do some check, coz we're claiming to be a safe wrapper and with no check we have undefined behaviour in safe code and that's not safe.

The resize should be very cheap --- get_num_atoms() is cached with lazy-init, so it's only a syscall the first time read() or get_num_atoms() is called on a trajectory. With the included iterator as written, later iterations will just compare the length of the frame vector to the cached u32 and then do nothing.

Reply:
Are you worried about the frame.resize() call or the trajectory.get_num_atoms() call?

Reply:
I actually forgot that `atom_num` is cached ;)

What do you think about raising an error if `get_num_atoms() != frame.len()`. I think this would be cleaner. If someone really passes a frame of incorrect size, there is very likely a bug in the piece of code that calls this function. Resizing just "hides" a bug in the users code.

Reply:
I hear what you're saying, but the advantage of resizing for the user is that its more ergonomic I guess, and it follows expectations that the standard library sets. I can do a `Vec::new() `and be confident i'll get the right answer, and if I find later that I'm spending lots of time allocating I can change it to `Vec::with_capacity()`, and I think that expectation would be worth maintaining here.

But yeah, I'd be happy with an error, as long as `frame.len()` points to the actual size of the vector. At the moment it points to `Frame.num_atoms`, and `Frame.coords` and `Frame.num_atoms` are both pub, so a user could change one and get UB. I think the best way forward there would just be to remove the num_atoms field, and have len look directly at the length of coords. We'd also need to document that requirement because its either not in the docs or I missed it.

It's getting late here so I'll pick this up tomorrow :)

Reply:
> I hear what you're saying, but the advantage of resizing for the user is that its more ergonomic I guess, and it follows expectations that the standard library sets. I can do a `Vec::new() `and be confident i'll get the right answer, and if I find later that I'm spending lots of time allocating I can change it to `Vec::with_capacity()`, and I think that expectation would be worth maintaining here.

But at this point, `with_capacity` has already been called with a wrong (too small) capacity. I think resizing would be equivalent to rust resizing vectors for you once you try to put something out of its bounds. i.e:

```Rust
let mut x = vec![0, 1, 2];
x[3] = 3;  // Does not resize x for you but errors.
```

Reply:
Ok yep, you've convinced me. But I think we should change `with_capacity` to `with_len`, so there's no ambiguity 🙂